### PR TITLE
Fix coroutine members beeing destroyed too late

### PR DESCRIPTION
**Pull request by jvolmer**
Before, a finished coroutine including its members were only destroyed after the garbage collection ran on the corresponding thead registry. This could potentially result in a deadlock because the coroutine can have aquired a lock which is only released after the coroutine is destroyed.
This makes it obvious that the coroutine and the promise inside the registry have different livetimes and should therefore be different objects. So instead of the thread registry having a list of coroutines as promises, it has now a list of its own promise objects. When a coroutine is created, it adds a promise object to the registry.

### CODE REVIEW

File: lib/Async/Registry/promise.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
+#include "promise.h"
```
Comment (jbajic) [inline lib/Async/Registry/promise.cpp:1]: Copyright header is missing

File: tests/Async/Registry/ThreadRegistryTest.cpp [hunk 2]
```diff
@@ -33,252 +33,192 @@ using namespace arangodb::async_registry;
...
 //   auto registry = ThreadRegistry::make(std::make_shared<Metrics>());
 
 //   std::jthread([&]() {
-//     auto promise = MyTestPromise{1};
-
-//     EXPECT_DEATH(registry->add(&promise), "Assertion failed");
+//     EXPECT_DEATH(registry->add_promise(),
+//                  "Assertion failed");
 //   });
 // }
```
Comment (jbajic) [inline tests/Async/Registry/ThreadRegistryTest.cpp:68]: Why can't we just delete this?

File: lib/Async/Registry/promise.cpp (hunk 1 above)
Comment (jvolmer) [reply lib/Async/Registry/promise.cpp:1]: Well spottet!

File: arangod/AsyncRegistryServer/Feature.h [hunk 3]
```diff
@@ -47,6 +47,8 @@ class Feature final : public application_features::ApplicationFeature {
     // startsAfter<Bla, Server>();
   }
 
+  ~Feature() { registry.set_metrics(std::make_shared<Metrics>()); }
```
Comment (maierlars) [inline arangod/AsyncRegistryServer/Feature.h:50]: why is this required? 🤔

**Review by maierlars — APPROVED** [review]

**Review by jbajic — APPROVED** [review]
